php/libraries/FilesUploadHandler.class.inc

```diff
@@ -0,0 +1,110 @@
+ * interface were created in order to encapsulate the usual things that happen
+ * when a set of files is uploaded to a server, namely:
+ *
+ * 1. Validation of all uploaded files
+ * 2. Transfer of all files to a designated (permanent) location on the
+ *    server.
+ * 3. Actual processing of the uploaded files (e.g. update of database
+ *    tables, etc...)
+ *
+ * @category Main
+ * @package  Main
+ * @author   Nicolas Brossard <nicolasbrossard.mni@gmail.com>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+interface FilesUploadHandler
+{
+    /**
+     * Returns the properties associated to each file uploaded on the
+     * server in an array identical to $_FILES.
+     *
+     * @return array the properties for each file uploaded (one array for
+     *         each uploaded file, see $_FILES).
+     */
+    public function getFiles() : array;
+
+    /**
+     * Validates the set of uploaded files. If a given file is invalid, the
+     * validation message associated to it will be stored in the array returned
+     * by getErrors().
+     *
+     * @return bool true if the set of files is valid and should be processed, false
+     *         otherwise.
+     */
+    public function isValid() : bool;
+
+    /**
+     * Returns any validation or processing errors associated to each uploaded file.
+     *
+     * @return array list of error messages for each uploaded file.
+     */
+    public function getErrors() : array;
+
+    /**
+     * Main entry point for classes implementing this interface. Processes all
+     * uploaded files.
+     *
+     * @return nothing.
+     */
+    public function processAll();
+
+    /**
+     * Gets the new path for a file that was uploaded to a temporary location on
+     * the server. This function only computes the new path and does not actually
+     * move the file.
+     *
+     * @param string $fileName temporary path on the server for a given uploaded
+     *               file.
+     *
+     * @return SplFileInfo all the information regarding the new file location.
+     */
+    public function getNewLocation(string $fileName) : SplFileInfo;
+
+    /**
+     * This function is called each time a file is moved to its permanent location
+     * on the server. This is where the actual processing of the uploaded file should
+     * take place.
+     *
+     * @param string $oldPath  temporary path on the server for a given uploaded
+     *               file.
+     * @param string $newPath  new (permanent) path for the uploaded file. If
+     *               the file could not actually be moved to this new location, the
+     *               value of this parameter will be null.
+     * @param string $errorMsg the error that occured during the move operation. If
+     *               the file was successfully moved to its new location, the value
+     *               for this parameter will be null.
+     *
+     * @return nothing.
```

Comment:
`NULL`

Reply:
What does PHPCS say it should be? I'm surprised it passed travis with "nothing."

Reply:
Yes, it passes Travis and phpcs does not complain. I'll replace with @return NULL if that is the convention.

Reply:
**@return void**  should be used to indicate the function doesn't return a value. `Null` indicate that the function return the __null__ value.

Reply:
Docs says: 

> If the return is omitted the value NULL will be returned. 

http://php.net/manual/en/functions.returning-values.php

Reply:
I would prefer void since it reminds me of C++ `void main (void)`

Reply:
I prefer void too but since the doc says that a void method returns NULL, I think that I have to use "@return NULL"...

Reply:
Ok for **null**, unlike C, php function always return _something_.